src/compress.c

```diff
@@ -0,0 +1,135 @@
+      return H5I_INVALID_HID;
+    }
+#ifdef HAVE_COMPRESSION
+    char *version = (char *) malloc(sizeof(char) * 512);
+    if (version == NULL) {
+      fprintf(stderr, "malloc for version failed\n");
+      free(version);
+      return H5I_INVALID_HID;
+    }
+    char *date = (char *) malloc(sizeof(char) * 512);
+    if (date == NULL) {
+      fprintf(stderr, "malloc for date failed\n");
+      free(version);
+      free(date);
+      return H5I_INVALID_HID;
+    }
+    cd_values[0] = 0;
+    cd_values[1] = 0;                 // compression_level;
+    cd_values[2] = 0;                 // shuffle;
+    cd_values[3] = 0;                 // blocksize
+    cd_values[4] = compression_level; // unused
+    cd_values[5] = shuffle;           // unused
+    cd_values[6] = blosc_compname_to_compcode(compressor);
+    if (H5Pset_filter(dcpl, FILTER_BLOSC, H5Z_FLAG_OPTIONAL, 7, cd_values) <
+        0) {
+      fprintf(stderr, "Error in H5Pset_filter\n");
+      return H5I_INVALID_HID;
+    }
+    if (register_blosc(&version, &date) < 0) {
+      fprintf(stderr, "Error in register_blosc\n");
+      free(version);
+      free(date);
+      return H5I_INVALID_HID;
+    } else {
+      printf("Blosc version info: %s (%s)\n", version, date);
+    }
+    free(version);
+    free(date);
+#endif
+  }
+  return dcpl;
+}
+
+int compress_frame(framebuffer *fb,
+                   unsigned int compression_level,
+                   unsigned int shuffle,
+                   char *compressor,
+                   size_t blocksize,
+                   int numinternalthreads)
+{
+  int bufsize = (fb->mq1_header->pixel_depth[1] - '0') * 10 +
+                (fb->mq1_header->pixel_depth[2] - '0');
+  bufsize = bufsize / 8;
+
+  int detx = (int) *(fb->mq1_header->det_x);
+  int dety = (int) *(fb->mq1_header->det_y);
+
+  size_t nbytes = dety * detx * bufsize;
+#ifdef HAVE_COMPRESSION
+  size_t destsize = nbytes + BLOSC_MAX_OVERHEAD;
+  void *dest      = malloc(destsize);
+  if (!dest) {
+    fprintf(stderr, "Error in malloc for dest\n");
+    return -1;
+  }
+
+  cbytes = blosc_compress_ctx(compression_level, shuffle, bufsize, nbytes,
+                              fb->data, dest, destsize, compressor, blocksize,
+                              numinternalthreads);
+
+  if (cbytes < 0) {
+    fprintf(stderr, "Error in blosc_compress\n");
+    free(dest);
+    return -1;
+  }
+  if (cbytes == 0) {
+    fprintf(stderr, "Blosc returned 0 bytes (uncompressible?). Forcing "
+                    "fallback to uncompressed write.\n");
+    cbytes = nbytes;
+    return cbytes;
+  }
+  // for showing compression ratio in each frame, profiling purposes
+  // if (cbytes != 0) {
+  //  printf("compression: %ld -> %d (%.1fx)\n", nbytes, cbytes, (1. * nbytes) /
+  //  cbytes);
+  //}
+
+  free(fb->data);
+  fb->data = malloc(destsize);
+  if (!fb->data) {
+    fprintf(stderr, "Error in malloc for fb->data\n");
+    free(dest);
+    return -1;
+  }
```

Comment:
Will the data in the original `fb->data` need to be accessed when this `malloc` fails? This free the original data there first and if `fb->data` needs to be accessed later, it will cause problem.